Optimize immutable.HashSet builder (inspired by the 2.13 implementation)

mkeskells opened the pull request:
The builder allows `TrieHashSet` to be mutated in place until it is visible this reduces the memory churn while building `Set`s.

Benchmark results: https://gist.github.com/mkeskells/7a4b80bb358f18afe541e0ce737aa691

mkeskells committed:
add a builder benchmark

retronym commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -1066,5 +1065,280 @@ object HashSet extends ImmutableSetFactory[HashSet] {
+    /* Nodes in the tree are either regular HashSet1, HashTrieSet, HashSetCollision1, or a mutable HashTrieSet
+      mutable HashTrieSet nodes are designated by having size == -1
+
+      mutable HashTrieSet nodes can have child nodes that are ummuable
```
"mutable"

mkeskells replied:
updated

retronym commented:
```
scala> import collection.immutable.HashSet; val b = HashSet.newBuilder[Int]; b ++= HashSet(1 to 10 : _*); b ++= HashSet(20 to 30: _*)
java.lang.AssertionError: assertion failed
  at scala.Predef$.assert(Predef.scala:208)
  at scala.collection.immutable.HashSet$HashSetBuilder.compressedIndex(HashSet.scala:1250)
  at scala.collection.immutable.HashSet$HashSetBuilder.trieIndex(HashSet.scala:1256)
  at scala.collection.immutable.HashSet$HashSetBuilder.addToTrieHashSet(HashSet.scala:1294)
  at scala.collection.immutable.HashSet$HashSetBuilder.addHashSet(HashSet.scala:1262)
  at scala.collection.immutable.HashSet$HashSetBuilder.$plus$plus$eq(HashSet.scala:1183)
  at scala.collection.immutable.HashSet$HashSetBuilder.$plus$plus$eq(HashSet.scala:1071)
```

mkeskells commented:
benchmark from baseline
https://gist.github.com/mkeskells/7a4b80bb358f18afe541e0ce737aa691

mkeskells commented:
removed the offending assert

retronym commented:
```
scala> case class C(i: Int) { override def hashCode = i % 1024 }
defined class C

scala> val b = collection.immutable.HashSet.newBuilder[C]; for (i <- 1 to 16) { b ++= {val b = collection.immutable.HashSet.newBuilder[C]; for (i <- 1 to 128) b += C(scala.util.Random.nextInt); b.result()}}; b.result()
java.lang.ArrayIndexOutOfBoundsException: 2
  at scala.collection.immutable.HashSet$HashSetBuilder.addToLeafHashSet(HashSet.scala:1441)
  at scala.collection.immutable.HashSet$HashSetBuilder.addHashSet(HashSet.scala:1372)
  at scala.collection.immutable.HashSet$HashSetBuilder.addToTrieHashSet(HashSet.scala:1412)
  at scala.collection.immutable.HashSet$HashSetBuilder.addHashSet(HashSet.scala:1373)
  at scala.collection.immutable.HashSet$HashSetBuilder.addToTrieHashSet(HashSet.scala:1412)
  at scala.collection.immutable.HashSet$HashSetBuilder.addHashSet(HashSet.scala:1373)
  at scala.collection.immutable.HashSet$HashSetBuilder.$plus$plus$eq(HashSet.scala:1295)
  at scala.collection.immutable.HashSet$HashSetBuilder.$plus$plus$eq(HashSet.scala:1183)
  at .$anonfun$new$1(<console>:13)
  at .$anonfun$new$1$adapted(<console>:13)
  at scala.collection.immutable.Range.foreach(Range.scala:158)
  ... 28 elided
```

retronym commented:
I'm hesitant to include this in 2.12.11 because of the closeness of the release and the size of this change. I'd be comfortable merging it straight after the release and then reaching out to folks to start testing nightlies. That could also include the other big backport PRs you have in progress like the `RedBlackTree` change.

mkeskells commented:
> I'm hesitant to include this in 2.12.11 because of the closeness of the release and the size of this change. I'd be comfortable merging it straight after the release and then reaching out to folks to start testing nightlies. That could also include the other big backport PRs you have in progress like the RedBlackTree change.

@retronym thats fine. For the work that inspired this we can take a private build from the 2.12. head and test it internally as part of that review process.

mkeskells committed:
Create a HashSet builder with limited mutability

The builder allows TrieHashSet to be mutated in place until it is visible
this reduces the memory churn while building Sets

don't allow build to emit an empty collection unless it the canonical one
i.e. (newBuilder() ++= new HashHet()).result() should be EmptyHashSet

mkeskells commented:
squashed

mkeskells committed:
Add a releaseFence before the builder publishes result
